### ISSUE TITLE
Doesn't render unicode characters?

### ISSUE THREAD

**Issue by ahmetkca** [asciinema/agg#28]
I have a cli application with spinner and I also use some unicode characters such as “🤖” (U+1F916).
They show  up as outline rectangles instead. I have also tried to use Nerd Font by specify custom `--font-dir` and `--font-family`

**Comment by ahmetkca** [comment]
I have tried `resvg` renderer instead of `fontdue` and it was able to render “⠸“ and it didn't render anything for “🤖”.

**Comment by sickill** [comment]
Can you provide example .cast file so we can try to reproduce the problem? Also, what OS?

**Comment by Zhengqbbb** [comment]
maybe can try add option `--renderer "resvg"`

**Comment by charlesnicholson** [comment]
I just hit this too on macOS 13.2.1 on a 2022 M2 MacBook Pro:

Expected:
![improg-expected](https://user-images.githubusercontent.com/3010295/226628063-e770f8be-8a41-4a22-80ae-99bec01c5da4.png)

Actual (no options):
![improg-bad-agg](https://user-images.githubusercontent.com/3010295/226628111-ec862754-189b-4fef-9dc1-c79422785504.gif)

Actual (`--render "resvg"`):
![improg-bad-agg-resvg](https://user-images.githubusercontent.com/3010295/226628198-ea3a7e9c-7933-464c-8f26-0907996a81de.gif)

I'm happy to share the cast file but GitHub doesn't allow the ".cast" file type as an attachment; how would you like to receive it? Alternately, if it's easier, my terminal app is trivial to reproduce, just clone and run `make` and then `./build/demo`: https://github.com/charlesnicholson/improg. It has some whitespace bugs on Linux and Windows but at least the emoji do render :)

**Comment by charlesnicholson** [comment]
I just discovered the `--font-family` flag, and they fix the display of the unicode characters in my prompt, but the 2-column emoji are all still broken. This happens with and without the `--renderer resvg` flag.

```
./agg-aarch64-apple-darwin --renderer resvg --font-family "Hack Nerd Font" improg-bad.cast improg.gif
```

**Comment by sickill** [comment]
I've just released 1.4.0 with much improved font loading and rendering behaviour. Please give it a try. Also see new [Fonts](https://github.com/asciinema/agg#fonts) section in README.

![issue-28](https://user-images.githubusercontent.com/17589/228838788-ea918380-fb01-4257-9d6a-a5631d649f3e.gif)

**Comment by charlesnicholson** [comment]
Thanks for the quick update! Unfortunately I'm not seeing any improvements for the various emoji (U+1F6..). With agg built from the 1.4.0 tag:

`../agg/target/release/agg improg.cast improg.gif --font-family "MesloLGS NF" --renderer resvg`
![improg resvg](https://user-images.githubusercontent.com/3010295/228848948-bbe5e1b8-1949-4c77-993f-68cfb69a1ca9.gif)

`../agg/target/release/agg improg.cast improg.gif --font-family "MesloLGS NF"`
![improg](https://user-images.githubusercontent.com/3010295/228849150-36d037ac-eb4b-4fd0-be93-559638ce5cfc.gif)

Here's a screen grab from my terminal as the demo is running:
<img width="1008" alt="Screenshot 2023-03-30 at 9 21 54 AM" src="https://user-images.githubusercontent.com/3010295/228849670-04a94588-7e1b-40d2-89df-5b191bab72ec.png">

Here's the cast file I'm using.
[improg.cast.zip](https://github.com/asciinema/agg/files/11111978/improg.cast.zip)

Perhaps it has something to do with my font? I'm using a nerd-patched Meno: https://github.com/ryanoasis/nerd-fonts/blob/master/patched-fonts/Meslo/M/Regular/complete/Meslo%20LG%20M%20Regular%20Nerd%20Font%20Complete.ttf

**Comment by sickill** [comment]
@charlesnicholson See [Fonts](https://github.com/asciinema/agg#fonts) section in README for info on emoji support. Without Noto Emoji font on your system you won't get any emoji rendered.

**Comment by sickill** [comment]
As for color emojis - agg uses fontdue library in its default renderer backend and this lib doesn't yet support color fonts. See https://github.com/mooman219/fontdue/issues/47 . Once fontdue adds support for it, e.g. for Noto Color Emoji as mentioned in that issue, we'll have it in agg.

We also have this alternative SVG based renderer - resvg - but similarly, it can't deal with color emoji properly yet. See https://github.com/RazrFalcon/resvg/issues/485#issuecomment-982854419

In other words, you could open a ticket but I don't plan to look into that until either of the libs gets better support for it, or until we write another renderer using a different text rendering library. So it's not really going to happen soon.

**Comment by ku1ik** [comment]
I just released agg 1.5.0 with improved support for emoji, including rendering color ones :tada: 

https://github.com/asciinema/agg/releases/tag/v1.5.0
https://docs.asciinema.org/manual/agg/usage/#emoji

If you have a chance to give it a shot then let me know how it performs now. Thanks!

**Comment by ku1ik** [comment]
Hi! A quick follow-up on this old issue: agg 1.8.0 is out and includes a new default renderer with improved Unicode and emoji fallback. It also bundles Noto Emoji as a fallback font, so characters like 🤖 should now work better out of the box.

Braille-style symbols should also have a better chance of rendering correctly through the new fallback path, though I’d still appreciate fresh reports if any specific characters are still missing.

This is how your recording looks now when rendered on macOS with agg 1.8.0:

<img width="1872" height="1092" alt="Image" src="https://github.com/user-attachments/assets/e4c877de-3cef-48bc-92da-8f15e21f3f4c" />

@charlesnicholson so, is asciinema there yet? :)
